enhancements/storage/vsphere-configurable-maximum-allowed-number-of-block-volumes-per-node.md

```diff
@@ -0,0 +1,229 @@
+current driver, which currently relies on a static limit that can not be changed by cluster administrators.
+
+## Summary
+
+The vSphere CSI driver for vSphere version 7 uses a constant to determine the maximum number of block volumes that can
+be attached to a single node. This limit is inflcuened by the number of SCSI controllers available on the node.
+By default, a node can have up to four SCSI controllers, each supporting up to 15 devices, allowing for a maximum of 60
+volumes per node (59 + root volume).
+
+However, vSphere version 8 increased the maximum number of volumes per node to 256 (255 + root volume). This enhancement
+aims to leverage this increased limit and provide administrators with finer-grained control over volume allocation
+allowing them to configure the maximum number of block volumes that can be attached to a single node.
+
+Details about configuration maximums: https://configmax.broadcom.com/guest?vmwareproduct=vSphere&release=vSphere%208.0&categories=3-0
+Volume limit configuration for vSphere storage plug-in: https://techdocs.broadcom.com/us/en/vmware-cis/vsphere/container-storage-plugin/3-0/getting-started-with-vmware-vsphere-container-storage-plug-in-3-0/vsphere-container-storage-plug-in-concepts/configuration-maximums-for-vsphere-container-storage-plug-in.html
+
+## Motivation
+
+### User Stories
+
+- As a vSphere administrator, I want to configure the maximum number of volumes that can be attached to a node, so that
+  I can optimize resource utilization and prevent oversubscription.
+- As a cluster administrator, I want to ensure that the vSphere CSI driver operates within the limits imposed by the 
+  underlying vSphere infrastructure.
+
+### Goals
+
+- Provide administrators with control over volume allocation limit on vSphere nodes.
+- Improve resource utilization and prevent oversubscription.
+- Ensure compatibility with existing vSphere infrastructure limitations.
+- Maintain backward compatibility with existing deployments.
+
+### Non-Goals
+
+- Support heterogeneous environments with different ESXi versions on nodes that form OpenShift cluster.
+- Dynamically adjust the limit based on real-time resource usage.
+- Implement per-namespace or per-workload volume limits.
+- Modify the underlying vSphere VM configuration.
+
+## Proposal
+
+1. Driver Feature State Switch (FSS):
+
+   - Use FSS (`max-pvscsi-targets-per-vm`) of the vSphere driver to control the activation of the maximum volume limit
+     functionality.
+   - No changes needed, the feature is enabled by default.
+
+2. API for Maximum Volume Limit:
+
+   - Introduce a new field `spec.driverConfig.vSphere.maxAllowedBlockVolumesPerNode` in ClusterCSIDriver API to allow
+     administrators to configure the desired maximum number of volumes per node.
+   - This field should default to the current maximum limit of 59 volumes per node for vSphere 7.
```

Comment:
Just to be clear: in my 4.18 cluster,  I don't have `maxAllowedBlockVolumesPerNode` field. When I upgrade to 4.19, the field appears and will have 59 as the value?

Reply:
Nope, I need to rewrite this - we decided not to default to 59 in the API and leave the default to be decided by operator. So after upgrading from 4.18, assuming we keep it tech preview in 4.19, nothing happens - we already set `MAX_VOLUMES_PER_NODE` in our controller and node deployment assets so no change there.

Then if users enable the tech preview the field will not show up (unset value is omitted from serialization and operator can set any default, we'll keep this at 59 for now). Then users can set `maxAllowedBlockVolumesPerNode` which will be reflected to deployments by operator. If the value is incorrect (e.g. users run vsphere 7 and set value above 59) we degrade.